### PR TITLE
[recovery] support linux 4.4 configfs based ub_gadget mode

### PR DESCRIPTION
**Pull request by Thaodan**
Ported the changes from hybris-boot to here.
However should the manufactor be Mer or Jolla?

### CODE REVIEW

**Comment by krnlyng** [comment]
@thaodan, looks nice but i think we should remove the mass storage parts if USB_FUNCTIONS is always rndis

**Comment by krnlyng** [comment]
write /sys/class/android_usb/android0/iManufacturer Jolla

should stay the same i think

**Comment by Thaodan** [comment]
Fixed, that. Should hybris-should support debug init like hybris-boot?

**Comment by krnlyng** [comment]
@Thaodan thanks!

This is the recovery so i don't think it needs extra debugging because it is already a debugging tool.

Or what do you mean?

**Comment by Thaodan** [comment]
I mean support being enabled along side boot as hybris-boot when used as init-debug.

File: recovery-init [hunk 1]
```diff
@@ -34,39 +104,45 @@ mkdir /dev/pts
...
+ipaddr add 192.168.2.15/24 broadcast 192.168.2.255 dev $USB_IFACE label rndis0:0
+
 udhcpd
 
 echo V > /dev/watchdog
 
-yamui -t "RECOVERY: Connect USB cable and open telnet to address 10.42.66.66" &
+
+usb_info "Jolla Debug telnet on $USB_IFACE $LOCAL_IP - also running udhcpd"
+yamui -t "RECOVERY: telnet on $USB_IFACE $LOCAL_IP" &
```
Comment (mkosola) [inline recovery-init]: Please keep yamui info same as it was before -> "RECOVERY: Connect USB cable and open telnet to address $USB_IFACE $LOCAL_IP"

**Comment by Thaodan** [comment]
Fixed

File: recovery-init [hunk 2]
```diff
@@ -24,6 +24,76 @@
...
+{
+	echo -n "$2" > $1
+}
+
+# This sets up the USB with whatever USB_FUNCTIONS are set to via configfs
+usb_setup_configfs() {
+    G_USB_ISERIAL=$GADGET_DIR/g1/strings/0x409/serialnumber
+    mkdir $GADGET_DIR/g1
+    write $GADGET_DIR/g1/idVendor                   "0x18D1"
+    write $GADGET_DIR/g1/idProduct                  "0xD001"
```
Comment (mkosola) [inline recovery-init]: These has also changed, please correct

write /sys/class/android_usb/android0/idVendor 2931
write /sys/class/android_usb/android0/idProduct 0A06

File: recovery-init [hunk 3]
```diff
@@ -24,6 +24,76 @@
...
+
+# This sets up the USB with whatever USB_FUNCTIONS are set to via android_usb
+usb_setup_android_usb() {
+    write $ANDROID_USB/enable          0
+    write $ANDROID_USB/functions       ""
+    write $ANDROID_USB/enable          1
+    usleep 500000 # 0.5 delay to attempt to remove rndis function
+    write $ANDROID_USB/enable          0
+    write $ANDROID_USB/idVendor        18D1
+    write $ANDROID_USB/idProduct       D001
```
Comment (mkosola) [inline recovery-init]: Same here:

write /sys/class/android_usb/android0/idVendor 2931
write /sys/class/android_usb/android0/idProduct 0A06

**Comment by Thaodan** [comment]
Changed as suggested.

File: recovery-init [hunk 4]
```diff
@@ -24,6 +24,76 @@
...
+
+# This sets up the USB with whatever USB_FUNCTIONS are set to via configfs
+usb_setup_configfs() {
+    G_USB_ISERIAL=$GADGET_DIR/g1/strings/0x409/serialnumber
+    mkdir $GADGET_DIR/g1
+    write $GADGET_DIR/g1/idVendor                   "0x2931"
+    write $GADGET_DIR/g1/idProduct                  "0x0A06"
+    mkdir $GADGET_DIR/g1/strings/0x409
+    write $GADGET_DIR/g1/strings/0x409/serialnumber "$1"
+    write $GADGET_DIR/g1/strings/0x409/manufacturer "Jolla Boat Loader"
```
Comment (mkosola) [inline recovery-init]: Ah sorry, missed this perviously... Still small change, could you leave manufacturer as Jolla

File: recovery-init [hunk 5]
```diff
@@ -24,6 +24,76 @@
...
+# This sets up the USB with whatever USB_FUNCTIONS are set to via android_usb
+usb_setup_android_usb() {
+    write $ANDROID_USB/enable          0
+    write $ANDROID_USB/functions       ""
+    write $ANDROID_USB/enable          1
+    usleep 500000 # 0.5 delay to attempt to remove rndis function
+    write $ANDROID_USB/enable          0
+    write $ANDROID_USB/idVendor        2931
+    write $ANDROID_USB/idProduct       0A06
+    write $ANDROID_USB/iManufacturer   "Jolla Boat Loader"
```
Comment (mkosola) [inline recovery-init]: Jolla also here

**Comment by Thaodan** [comment]
Np, done.

**Comment by mkosola** [comment]
Thank you, looks good now, I'll test this on Monday and merge after that.

File: recovery-init [hunk 6]
```diff
@@ -34,39 +104,44 @@ mkdir /dev/pts
...
-write /sys/class/android_usb/android0/iSerial 123456789
-write /sys/class/android_usb/android0/enable 1
+
+
+usb_setup "Jolla Debug setting up"
 
 ip link set rndis0 up
-ipaddr add 10.42.66.66/29 broadcast 10.42.66.255 dev rndis0
-ipaddr add 192.168.2.15/24 broadcast 192.168.2.255 dev rndis0 label rndis0:0
+if ipaddr add 10.42.66.66/29 broadcast 10.42.66.255 dev rndis0 ; then
```
Comment (mkosola) [inline recovery-init]: Noticed still one functional difference compared to old release when testing. Could you add following line after then and before USB_IFACE=rndis0:
ipaddr add 192.168.2.15/24 broadcast 192.168.2.255 dev rndis0 label rndis0:0

Comment (Thaodan) [reply recovery-init]: I have already removed that line. Please pull a gain.

Comment (Thaodan) [reply recovery-init]: Fixed that but used $USB_IFACE instead of static rndis0.

Comment (mkosola) [reply recovery-init]: `ip link set rndis0 up
if ipaddr add 10.42.66.66/29 broadcast 10.42.66.255 dev rndis0 ; then
    ipaddr add 192.168.2.15/24 broadcast 192.168.2.255 dev rndis0 label rndis0:0
    USB_IFACE=rndis0
elif ipaddr add 10.42.66.66/29 broadcast 10.42.66.255 dev usb0 ; then`

Comment (mkosola) [reply recovery-init]: Sorry, meant like that ^

**Pull request merged**